File: python/src/context.py
from .api.endpoint.CREATE_Installation import CREATE_Installation
from .api.endpoint.CREATE_DeviceServer import CREATE_DeviceServer
from .api.endpoint.CREATE_SessionServer import CREATE_SessionServer

from .api.model.DeviceServer import DeviceServer
from .api.model.Installation import Installation
from .api.model.SessionServer import SessionServer

from .client import Client
from .signing import generate_rsa_key_pair

from .wirespec import Serialization

class Context:
    serialization = Serialization()
    client = Client(serialization)

    private_key_pem, public_key_pem = generate_rsa_key_pair()



    def __init__(self, api_key: str, service_name: str):
        self.api_key = api_key
        self.service_name = service_name
        installation = self.create_installation(service_name)
        self.server_public_key = installation.ServerPublicKey.server_public_key
        device_server = self.create_device_server(service_name, api_key, installation.Token.token)
        self.device_id = device_server.Id.id
        session_server = self.create_session_server(service_name, api_key, installation.Token.token)
        self.session_id = session_server.Id.id
        self.session_token = session_server.Token.token
        self.user_id = session_server.UserPerson.id

    def create_installation(self, service_name:str) -> Installation:
        body = Installation(
            client_public_key = self.public_key_pem
        )
        req = CREATE_Installation.Request(
            CacheControl = None,
            UserAgent = service_name,
            XBunqLanguage = None,
            XBunqRegion = None,
            XBunqClientRequestId = None,
            XBunqGeolocation = None,
            XBunqClientAuthentication = "",
            body = body
        )
        res = self.client.CREATE_Installation(req)
        match res:
            case CREATE_Installation.Response200(body=installation):
                return installation
            case _:
                raise Exception("Cannot create installation")

    def create_device_server(self, service_name:str, api_key:str, token:str) -> DeviceServer:
        body = DeviceServer(
            description = service_name,
            secret = api_key,
            permitted_ips = ["*"]
        )
        req = CREATE_DeviceServer.Request(
            CacheControl = None,
            UserAgent = service_name,
            XBunqLanguage = None,
            XBunqRegion = None,
            XBunqClientRequestId = None,
            XBunqGeolocation = None,
            XBunqClientAuthentication = token,
            body = body
        )
        match self.client.CREATE_DeviceServer(req):
            case CREATE_DeviceServer.Response200(body=device_server):
                return device_server
            case _:
                raise Exception("Cannot create device server")

    def create_session_server(self, service_name:str, api_key:str, token:str) -> SessionServer:
        body = SessionServer(
            secret = api_key,
        )
        req = CREATE_SessionServer.Request(
            CacheControl = None,
            UserAgent = service_name,
            XBunqLanguage = None,
            XBunqRegion = None,
            XBunqClientRequestId = None,
            XBunqGeolocation = None,
            XBunqClientAuthentication = token,
            body = body
        )
        match self.client.CREATE_SessionServer(req):
            case CREATE_SessionServer.Response200(body=server_session):
                return server_session
            case _:
                raise Exception("Cannot create device server")

File: java/lib/src/main/java/community/flock/bunq/java/Context.java
package community.flock.bunq.java;


import com.bunq.sdk.security.SecurityUtils;
import community.flock.bunq.java.context.InstallationContext;
import community.flock.wirespec.generated.java.CREATE_DeviceServerEndpoint;
import community.flock.wirespec.generated.java.CREATE_InstallationEndpoint;
import community.flock.wirespec.generated.java.DeviceServer;
import community.flock.wirespec.generated.java.Installation;

import java.security.KeyPair;
import java.security.PublicKey;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

/**
 * API context
 *
 * @param apiKey
 * @param serverName
 * @param serverPublicKey
 * @param keyPair
 * @param deviceId
 * @param sessionId
 * @param sessionToken
 * @param userId
 */
public record Context(
        String apiKey,
        String serverName,
        PublicKey serverPublicKey,
        KeyPair keyPair,
        int deviceId,
        int sessionId,
        String sessionToken,
        int userId
) {
}

public class ContextX {
    private final String apiKey;
    private final String serverName;
    private PublicKey serverPublicKey;
    private KeyPair keyPair;
    private final int deviceId;
    private final int sessionId;
    private String sessionToken;
    private final int userId;

    public ContextX(String apiKey,
                    String deviceDescription,
                    List<String> permittedIps,
                    String proxy) {
        this.apiKey = apiKey;
        this.serverName = deviceDescription;

        initialize(deviceDescription, permittedIps);


    }

    public ContextX create(String apiKey, String deviceDescription,
                           List<String> permittedIps, ContextInitializerHandler requestHandler) {
        String xBunqClientAuthentication = "bunq-client-authd";
        Installation body = new Installation("client-public-key");

        String userAgent = "Hallo from Flock bunq sdk";
        var responseCompletableFuture = requestHandler.cREATE_Installation(new CREATE_InstallationEndpoint.Request(
                        Optional.empty(),
                        userAgent,
                        Optional.empty(),
                        Optional.empty(),
                        Optional.empty(),
                        Optional.empty(),
                        xBunqClientAuthentication,
                        body

                )).thenApply(it ->
                        switch (it) {
                            case CREATE_InstallationEndpoint.ResponseCREATE_Installation400ResponseBody responseCREATEInstallation400ResponseBody ->
                                    null;
                            case CREATE_InstallationEndpoint.ResponseInstallationCreate responseInstallationCreate ->
                                    throw new RuntimeException("Whoopsie");
                        }
                )
                .thenApply(it -> {

                            DeviceServer deviceServer = new DeviceServer(
                                    "description",
                                    "secret",
                                    Optional.of(List.of(permittedIps.toArray(new String[0])))
                            );
                            return requestHandler.cREATE_DeviceServer(new CREATE_DeviceServerEndpoint.Request(
                                    Optional.empty(),
                                    userAgent,
                                    Optional.empty(),
                                    Optional.empty(),
                                    Optional.empty(),
                                    Optional.empty(),
                                    xBunqClientAuthentication,
                                    deviceServer

                            ));
                        }

                ).thenApply(it ->
                        null
                        );


        var responseCompletableFuture2 = requestHandler.cREATE_SessionServer(null/*TODO */);
    }

    private void initialize(String deviceDescription, List<String> permittedIps) {
        /* The calls below are order-sensitive: to initialize a Device Registration, we need an
         * Installation, and to initialize a Session we need a Device Registration. */
        initializeInstallation();
        initializeDeviceRegistration(deviceDescription, permittedIps);
        initializeSession();
    }

    /**
     * Create a new installation and store its data in an InstallationContext.
     */
    private void initializeInstallation() {
        this.keyPair = SecurityUtils.generateKeyPair();

        var publicKeyClientString = SecurityUtils.getPublicKeyFormattedString(this.keyPair.getPublic());
        this.sessionToken = installation.getSessionToken().getToken();
        this.serverPublicKey = SecurityUtils.createPublicKeyFromFormattedString(installation.getPublicKeyServer());
        this.serverPublicKey = installation.getPublicKeyServer();
        installationContext = new InstallationContext(installation, keyPairClient);
    }

    private static byte[] generateRequestBodyBytes(String publicKeyClientString) {
        HashMap<String, Object> installationRequestBody = new HashMap<>();
        installationRequestBody.put(FIELD_CLIENT_PUBLIC_KEY, publicKeyClientString);

        return gson.toJson(installationRequestBody).getBytes();
    }
}